src/app.py

```diff
@@ -1,62 +1,196 @@
         super().__init__(path)
+        # Extract field names, skipping the first DeletionFlag field
+        self.field_names = [field[0] for field in self.fields[1:]]
+        # Precompute Shapely geometries for all shapes
+        self.shapely_geometries = [
+            shapely_shape(self.shape(i).__geo_interface__) for i in range(len(self))
+        ]
 
     def geometry(self, i) -> list:
         return self.shape(i).points
-    
+
+    def shapely_geometry(self, i):
+        """
+        Return the Shapely geometry for the ith shape.
+        """
+        return self.shapely_geometries[i]
+
     def record(self, i) -> dict:
-        return super().record(i)
-    
+        """
+        Return a dictionary mapping field names to their corresponding values.
+        """
+        rec = super().record(i)
+        return dict(zip(self.field_names, rec))
+
     def __len__(self) -> int:
         return super().__len__()
 
-def main():
-    ### Config ###
-    plot = False # Set True of you want the catchment to be plotted
-    model = pyromb.RORB() # Select your hydrology model, either pyromb.RORB() or pyromb.WBNM()
+
+def main(
+    reach_path: str | None = None,
+    basin_path: str | None = None,
+    centroid_path: str | None = None,
+    confluence_path: str | None = None,
+    output_name: str | None = None,
+    plot: bool = False,
+    model: Any | None = None,
+) -> None:
+    """
+    Main function to build and process catchment data.
+
+    Parameters
+    ----------
+    reach_path : str
+        Path to the reaches shapefile.
+    basin_path : str
+        Path to the basins shapefile.
+    centroid_path : str
+        Path to the centroids shapefile.
+    confluence_path : str
+        Path to the confluences shapefile.
+    output_name : str
+        Name of the output file.
+    plot : bool
+        Whether to plot the catchment.
+    model : pyromb.Model
+        The hydrology model to use.
+    """
+    # Set default paths if not provided
+    reach_path = reach_path or REACH_PATH
+    basin_path = basin_path or BASIN_PATH
+    centroid_path = centroid_path or CENTROID_PATH
+    confluence_path = confluence_path or CONFUL_PATH
+    model = model or pyromb.RORB()
+    if isinstance(model, pyromb.RORB):
+        output_name = output_name or os.path.join(DIR, "../vector.catg")
+    else:
+        output_name = output_name or os.path.join(DIR, "../runfile.wbnm")
+    model = model or pyromb.RORB()
 
     ### Build Catchment Objects ###
-    # Vector layers 
-    reach_vector = SFVectorLayer(REACH_PATH)
-    basin_vector = SFVectorLayer(BASIN_PATH)
-    centroid_vector = SFVectorLayer(CENTROID_PATH)
-    confluence_vector = SFVectorLayer(CONFUL_PATH)
-    # Create the builder. 
+    # Vector layers
+    reach_vector = SFVectorLayer(reach_path)
+    basin_vector = SFVectorLayer(basin_path)
+    centroid_vector = SFVectorLayer(centroid_path)
+    confluence_vector = SFVectorLayer(confluence_path)
+
+    # Validate shapefile fields
```

Comment:
Sorry, I haven't coded a QGIS plugin before so just tried to follow the sequence of what was being called. 
I tried to avoid changing too many of your files to minimise things that I could break.
Shouldn't be too hard to move the validation into the library proper.

Reply:
All you need to know about the QGIS plugin is that its job is read and wrap the shapefiles in a consistent interface then pass to Pyromb which does all the heavy lifting. Pyromb is trying to be as independent as possible so I didnt want any other software doing too much.